Move Rule evaluation to Source

oxzi opened the pull request:
This is a first version to move the rule evaluation from Icinga Notifications into the source, starting with Icinga DB.

For an end user or another computer, the /process-event API endpoint was slightly modified. Two new HTTP request headers were introduced to tell which rules match and to share the state.

Another new /event-rules API endpoint allows querying all rules.

---

Other PR is Icinga/icingadb#998.

---

At the moment, there is no real support in Icinga Notifications Web. Thus, start by creating a rule through the current web interface and update the data in the relational database. Consider the following:

```
notifications=# select name, object_filter from rule;
-[ RECORD 1 ]-+------------------------------------------------------------------------------------------------
name          | hosts with 1
object_filter | select * from host where id = :host_id and environment_id = :environment_id and name like '%1%'
-[ RECORD 2 ]-+------------------------------------------------------------------------------------------------
name          | only services
object_filter | select * from service where id = :service_id and environment_id = :environment_id
-[ RECORD 3 ]-+------------------------------------------------------------------------------------------------
name          | fair dice
object_filter | select rand() as r having r >= 0.4
```

oxzi commented on internal/config/rule.go:
```diff
@@ -6,8 +6,36 @@ import (
+type SourceRulesInfo struct {
+	// Version is the version of the rules for the source.
+	//
+	// It is a monotonically increasing number that is updated whenever a rule is added, modified, or deleted.
+	// With each state change of the rules referenced by RuleIDs, the Version will always be incremented
+	// by 1, starting from 0. When there are no configured rules for the source, Version will be reset to 0.
```
> It is a monotonically increasing number that is updated whenever a rule is added, modified, or deleted.

Based on the implementation below, the internal deletion handler method does not touch the `updatedSources` map. So the `.Version` field will not be incremented on rule deletion.

yhabteab replied:
Unveiled a not yet merged bug :)!

yhabteab replied:
I was just thinking that a source without any rules shouldn't have any version set, even when the rules are deleted at runtime. However, since that isn't implemented correctly anyway (as you already found out) and @julianbrost lready pointed to its potential problems, you can remove that logic completely.

oxzi commented:
The following diff against Icinga Notification Web "works" if your source has the ID 1. It is based on @yhabteab's diff from above. It should apply on top of today's `main` branch, being at 6d193d1058ad153330a37b70fea4992b9683822a.

<details><summary>CLICK HERE FOR A FREE DOWNLOAD</summary>
<p>

```diff
diff --git a/application/controllers/EventRuleController.php b/application/controllers/EventRuleController.php
index 9d07d6b..6ffdf3c 100644
--- a/application/controllers/EventRuleController.php
+++ b/application/controllers/EventRuleController.php
@@ -12,7 +12,6 @@ use Icinga\Module\Notifications\Forms\EventRuleConfigElements\NotificationConfig
 use Icinga\Module\Notifications\Forms\EventRuleConfigForm;
 use Icinga\Module\Notifications\Forms\EventRuleForm;
 use Icinga\Module\Notifications\Model\Rule;
-use Icinga\Module\Notifications\Web\Control\SearchBar\ExtraTagSuggestions;
 use Icinga\Web\Notification;
 use Icinga\Web\Session;
 use ipl\Html\Form;
@@ -157,9 +156,6 @@ class EventRuleController extends CompatController
      */
     public function completeAction(): void
     {
-        $suggestions = new ExtraTagSuggestions();
-        $suggestions->forRequest($this->getServerRequest());
-        $this->getDocument()->add($suggestions);
     }
 
     /**
diff --git a/application/forms/EventRuleConfigForm.php b/application/forms/EventRuleConfigForm.php
index 34da1ac..2652535 100644
--- a/application/forms/EventRuleConfigForm.php
+++ b/application/forms/EventRuleConfigForm.php
@@ -291,6 +291,7 @@ class EventRuleConfigForm extends Form
             $db->insert('rule', [
                 'name'          => $this->getValue('name'),
                 'timeperiod_id' => null,
+                'source_id'     => 1, // TODO :^)
                 'object_filter' => $this->getValue('object_filter'),
                 'changed_at'    => (int) (new DateTime())->format("Uv"),
                 'deleted'       => 'n'
diff --git a/application/forms/SourceForm.php b/application/forms/SourceForm.php
index dcfbc63..a57f626 100644
--- a/application/forms/SourceForm.php
+++ b/application/forms/SourceForm.php
@@ -71,144 +71,48 @@ class SourceForm extends CompatForm
             || ($chosenType === null && $configuredType === null)
             || ($chosenType === null && $configuredType === Source::ICINGA_TYPE_NAME);
 
-        if ($showIcingaApiConfig) {
-            // TODO: Shouldn't be necessary: https://github.com/Icinga/ipl-html/issues/131
-            $this->clearPopulatedValue('type');
+        $this->getElement('icinga_or_other')->setValue('other');
 
-            $this->addElement(
-                'hidden',
-                'type',
-                [
-                    'required'  => true,
-                    'disabled'  => true,
-                    'value'     => Source::ICINGA_TYPE_NAME
-                ]
-            );
-            $this->addElement(
-                'text',
-                'icinga2_base_url',
-                [
-                    'required'  => true,
-                    'label'     => $this->translate('API URL')
-                ]
-            );
-            $this->addElement(
-                'text',
-                'icinga2_auth_user',
-                [
-                    'required'      => true,
-                    'label'         => $this->translate('API Username'),
-                    'autocomplete'  => 'off'
-                ]
-            );
-            $this->addElement(
-                'password',
-                'icinga2_auth_pass',
-                [
-                    'required'      => true,
-                    'label'         => $this->translate('API Password'),
-                    'autocomplete'  => 'new-password'
-                ]
-            );
-            $this->addElement(
-                'checkbox',
-                'icinga2_insecure_tls',
-                [
-                    'class'             => 'autosubmit',
-                    'label'             => $this->translate('Verify API Certificate'),
-                    'checkedValue'      => 'n',
-                    'uncheckedValue'    => 'y',
-                    'value'             => true
-                ]
-            );
-
-            /** @var CheckboxElement $insecureBox */
-            $insecureBox = $this->getElement('icinga2_insecure_tls');
-            if ($insecureBox->isChecked()) {
-                $this->addElement(
-                    'text',
-                    'icinga2_common_name',
-                    [
-                        'label'         => $this->translate('Common Name'),
-                        'description'   => $this->translate(
-                            'The CN of the API certificate. Only required if it differs from the FQDN of the API URL'
-                        )
-                    ]
-                );
-
-                $this->addElement(
-                    'textarea',
-                    'icinga2_ca_pem',
-                    [
-                        'cols'          => 64,
-                        'rows'          => 28,
-                        'required'      => true,
-                        'validators'    => [new X509CertValidator()],
-                        'label'         => $this->translate('CA Certificate'),
-                        'description'   => $this->translate('The certificate of the Icinga CA')
-                    ]
-                );
-            }
-        } else {
-            $this->getElement('icinga_or_other')->setValue('other');
-
-            $this->addElement(
-                'text',
-                'type',
-                [
-                    'required'      => true,
-                    'label'         => $this->translate('Type Name'),
-                    'validators'    => [new CallbackValidator(function (string $value, CallbackValidator $validator) {
-                        if ($value === Source::ICINGA_TYPE_NAME) {
-                            $validator->addMessage($this->translate('This name is reserved and cannot be used'));
-
-                            return false;
-                        }
-
-                        return true;
-                    })]
-                ]
-            );
-            $this->addElement(
-                'password',
-                'listener_password',
-                [
-                    'ignore'        => true,
-                    'required'      => $this->sourceId === null,
-                    'label'         => $this->sourceId !== null
-                        ? $this->translate('New Password')
-                        : $this->translate('Password'),
-                    'autocomplete'  => 'new-password',
-                    'validators'    => [['name' => 'StringLength', 'options' => ['min' => 16]]]
-                ]
-            );
-            $this->addElement(
-                'password',
-                'listener_password_dupe',
-                [
-                    'ignore'        => true,
-                    'required'      => $this->sourceId === null,
-                    'label'         => $this->translate('Repeat Password'),
-                    'autocomplete'  => 'new-password',
-                    'validators'    => [new CallbackValidator(function (string $value, CallbackValidator $validator) {
-                        if ($value !== $this->getValue('listener_password')) {
-                            $validator->addMessage($this->translate('Passwords do not match'));
-
-                            return false;
-                        }
-
-                        return true;
-                    })]
-                ]
-            );
-
-            // Preserves (some) entered data even if the user switches between types
-            $this->addElement('hidden', 'icinga2_base_url');
-            $this->addElement('hidden', 'icinga2_auth_user');
-            $this->addElement('hidden', 'icinga2_insecure_tls');
-            $this->addElement('hidden', 'icinga2_common_name');
-            $this->addElement('hidden', 'icinga2_ca_pem');
-        }
+        $this->addElement(
+            'text',
+            'type',
+            [
+                'required'      => true,
+                'label'         => $this->translate('Type Name'),
+            ]
+        );
+        $this->addElement(
+            'password',
+            'listener_password',
+            [
+                'ignore'        => true,
+                'required'      => $this->sourceId === null,
+                'label'         => $this->sourceId !== null
+                    ? $this->translate('New Password')
+                    : $this->translate('Password'),
+                'autocomplete'  => 'new-password',
+                'validators'    => [['name' => 'StringLength', 'options' => ['min' => 16]]]
+            ]
+        );
+        $this->addElement(
+            'password',
+            'listener_password_dupe',
+            [
+                'ignore'        => true,
+                'required'      => $this->sourceId === null,
+                'label'         => $this->translate('Repeat Password'),
+                'autocomplete'  => 'new-password',
+                'validators'    => [new CallbackValidator(function (string $value, CallbackValidator $validator) {
+                    if ($value !== $this->getValue('listener_password')) {
+                        $validator->addMessage($this->translate('Passwords do not match'));
+
+                        return false;
+                    }
+
+                    return true;
+                })]
+            ]
+        );
 
         $this->addElement(
             'submit',
@@ -376,12 +280,6 @@ class SourceForm extends CompatForm
         return [
             'name'                  => $source->name,
             'type'                  => $source->type,
-            'icinga2_base_url'      => $source->icinga2_base_url,
-            'icinga2_auth_user'     => $source->icinga2_auth_user,
-            'icinga2_auth_pass'     => $source->icinga2_auth_pass,
-            'icinga2_ca_pem'        => $source->icinga2_ca_pem,
-            'icinga2_common_name'   => $source->icinga2_common_name,
-            'icinga2_insecure_tls'  => $source->icinga2_insecure_tls
         ];
     }
 }
diff --git a/library/Notifications/Model/ExtraTag.php b/library/Notifications/Model/ExtraTag.php
deleted file mode 100644
index 07a27f8..0000000
--- a/library/Notifications/Model/ExtraTag.php
+++ /dev/null
@@ -1,28 +0,0 @@
-<?php
-
-/* Icinga Notifications Web | (c) 2023 Icinga GmbH | GPLv2 */
-
-namespace Icinga\Module\Notifications\Model;
-
-use Icinga\Module\Notifications\Model\Behavior\ObjectTags;
-use ipl\Orm\Behaviors;
-use ipl\Sql\Connection;
-
-class ExtraTag extends ObjectExtraTag
-{
-    /**
-     * @internal Don't use. This model acts only as relation target and is not supposed to be directly used as query
-     *           target. Use {@see ObjectExtraTag} instead.
-     */
-    public static function on(Connection $_)
-    {
-        throw new \LogicException('Documentation says: DO NOT USE. Can\'t you read?');
-    }
-
-    public function createBehaviors(Behaviors $behaviors): void
-    {
-        parent::createBehaviors($behaviors);
-
-        $behaviors->add(new ObjectTags());
-    }
-}
diff --git a/library/Notifications/Model/ObjectExtraTag.php b/library/Notifications/Model/ObjectExtraTag.php
deleted file mode 100644
index df7beb7..0000000
--- a/library/Notifications/Model/ObjectExtraTag.php
+++ /dev/null
@@ -1,52 +0,0 @@
-<?php
-
-/* Icinga Notifications Web | (c) 2023 Icinga GmbH | GPLv2 */
-
-namespace Icinga\Module\Notifications\Model;
-
-use ipl\Orm\Behavior\Binary;
-use ipl\Orm\Behaviors;
-use ipl\Orm\Model;
-use ipl\Orm\Query;
-use ipl\Orm\Relations;
-
-/**
- * ObjectExtraTag database model
- *
- * @property string $object_id
- * @property string $tag
- * @property string $value
- *
- * @property Query|Objects $object
- */
-class ObjectExtraTag extends Model
-{
-    public function getTableName(): string
-    {
-        return 'object_extra_tag';
-    }
-
-    public function getKeyName(): array
-    {
-        return ['object_id', 'tag'];
-    }
-
-    public function getColumns(): array
-    {
-        return [
-            'object_id',
-            'tag',
-            'value'
-        ];
-    }
-
-    public function createBehaviors(Behaviors $behaviors): void
-    {
-        $behaviors->add(new Binary(['object_id']));
-    }
-
-    public function createRelations(Relations $relations): void
-    {
-        $relations->belongsTo('object', Objects::class);
-    }
-}
diff --git a/library/Notifications/Model/Objects.php b/library/Notifications/Model/Objects.php
index c621056..3f37a27 100644
--- a/library/Notifications/Model/Objects.php
+++ b/library/Notifications/Model/Objects.php
@@ -25,8 +25,6 @@ use ipl\Orm\Relations;
  * @property Query|Event $event
  * @property Query|Incident $incident
  * @property Query|Tag $tag
- * @property Query|ObjectExtraTag $object_extra_tag
- * @property Query|ExtraTag $extra_tag
  * @property Query|Source $source
  * @property array<string, string> $id_tags
  */
@@ -81,10 +79,6 @@ class Objects extends Model
 
         $relations->hasMany('object_id_tag', ObjectIdTag::class);
         $relations->hasMany('tag', Tag::class);
-        $relations->hasMany('object_extra_tag', ObjectExtraTag::class)
-            ->setJoinType('LEFT');
-        $relations->hasMany('extra_tag', ExtraTag::class)
-            ->setJoinType('LEFT');
 
         $relations->belongsTo('source', Source::class)->setJoinType('LEFT');
     }
diff --git a/library/Notifications/Model/Rule.php b/library/Notifications/Model/Rule.php
index 8e69a01..00cd5f8 100644
--- a/library/Notifications/Model/Rule.php
+++ b/library/Notifications/Model/Rule.php
@@ -16,6 +16,7 @@ use ipl\Orm\Relations;
  * @property int $id
  * @property string $name
  * @property ?int $timeperiod_id
+ * @property ?int $source_id
  * @property ?string $object_filter
  * @property DateTime $changed_at
  * @property bool $deleted
@@ -41,6 +42,7 @@ class Rule extends Model
         return [
             'name',
             'timeperiod_id',
+            'source_id',
             'object_filter',
             'changed_at',
             'deleted'
@@ -51,6 +53,7 @@ class Rule extends Model
     {
         return [
             'name'          => t('Name'),
+            'source_id'     => t('Source ID'),
             'timeperiod_id' => t('Timeperiod ID'),
             'object_filter' => t('Object Filter'),
             'changed_at'    => t('Changed At')
diff --git a/library/Notifications/Model/Source.php b/library/Notifications/Model/Source.php
index 973bbb7..8558ace 100644
--- a/library/Notifications/Model/Source.php
+++ b/library/Notifications/Model/Source.php
@@ -19,12 +19,6 @@ use ipl\Web\Widget\Icon;
  * @property string $type Type identifier
  * @property string $name The user-defined name
  * @property ?string $listener_password_hash
- * @property ?string $icinga2_base_url
- * @property ?string $icinga2_auth_user
- * @property ?string $icinga2_auth_pass
- * @property ?string $icinga2_ca_pem
- * @property ?string $icinga2_common_name
- * @property string $icinga2_insecure_tls
  * @property DateTime $changed_at
  * @property bool $deleted
  *
@@ -33,7 +27,7 @@ use ipl\Web\Widget\Icon;
 class Source extends Model
 {
     /** @var string The type name used by Icinga sources */
-    public const ICINGA_TYPE_NAME = 'icinga2';
+    public const ICINGA_TYPE_NAME = 'icingadb';
 
     public function getTableName(): string
     {
@@ -51,12 +45,6 @@ class Source extends Model
             'type',
             'name',
             'listener_password_hash',
-            'icinga2_base_url',
-            'icinga2_auth_user',
-            'icinga2_auth_pass',
-            'icinga2_ca_pem',
-            'icinga2_common_name',
-            'icinga2_insecure_tls',
             'changed_at',
             'deleted'
         ];
diff --git a/library/Notifications/Web/Control/SearchBar/ExtraTagSuggestions.php b/library/Notifications/Web/Control/SearchBar/ExtraTagSuggestions.php
deleted file mode 100644
index 88f52c2..0000000
--- a/library/Notifications/Web/Control/SearchBar/ExtraTagSuggestions.php
+++ /dev/null
@@ -1,41 +0,0 @@
-<?php
-
-/* Icinga Notifications Web | (c) 2023 Icinga GmbH | GPLv2 */
-
-namespace Icinga\Module\Notifications\Web\Control\SearchBar;
-
-use Icinga\Module\Notifications\Common\Database;
-use Icinga\Module\Notifications\Model\ObjectExtraTag;
-use Icinga\Module\Notifications\Util\ObjectSuggestionsCursor;
-use ipl\Web\Control\SearchBar\Suggestions;
-use ipl\Stdlib\Filter;
-use Traversable;
-
-class ExtraTagSuggestions extends Suggestions
-{
-    protected function fetchValueSuggestions($column, $searchTerm, Filter\Chain $searchFilter)
-    {
-        yield;
-    }
-
-    protected function createQuickSearchFilter($searchTerm)
-    {
-        return Filter::any();
-    }
-
-    protected function fetchColumnSuggestions($searchTerm)
-    {
-        $searchColumns = (new ObjectSuggestionsCursor(
-            Database::get(),
-            (new ObjectExtraTag())::on(Database::get())
-                ->columns(['tag'])
-                ->assembleSelect()
-                ->distinct()
-        ));
-
-        // Object Extra Tags
-        foreach ($searchColumns as $column) {
-            yield $column->tag => $column->tag;
-        }
-    }
-}
diff --git a/library/Notifications/Web/Control/SearchBar/ObjectSuggestions.php b/library/Notifications/Web/Control/SearchBar/ObjectSuggestions.php
index 2a4d674..2bd1b82 100644
--- a/library/Notifications/Web/Control/SearchBar/ObjectSuggestions.php
+++ b/library/Notifications/Web/Control/SearchBar/ObjectSuggestions.php
@@ -6,7 +6,6 @@ namespace Icinga\Module\Notifications\Web\Control\SearchBar;
 
 use Icinga\Module\Notifications\Common\Auth;
 use Icinga\Module\Notifications\Common\Database;
-use Icinga\Module\Notifications\Model\ObjectExtraTag;
 use Icinga\Module\Notifications\Model\ObjectIdTag;
 use Icinga\Module\Notifications\Util\ObjectSuggestionsCursor;
 use ipl\Html\HtmlElement;
@@ -184,9 +183,9 @@ class ObjectSuggestions extends Suggestions
         }
 
         // Custom variables only after the columns are exhausted and there's actually a chance the user sees them
-        foreach ([new ObjectIdTag(), new ObjectExtraTag()] as $model) {
+        foreach ([new ObjectIdTag()] as $model) {
             $titleAdded = false;
-            /** @var ObjectIdTag|ObjectExtraTag $tag */
+            /** @var ObjectIdTag $tag */
             foreach ($this->queryTags($model, $searchTerm) as $tag) {
                 $isIdTag = $tag instanceof ObjectIdTag;
 
diff --git a/library/Notifications/Widget/Detail/IncidentDetail.php b/library/Notifications/Widget/Detail/IncidentDetail.php
index 46e6bed..a9c370b 100644
--- a/library/Notifications/Widget/Detail/IncidentDetail.php
+++ b/library/Notifications/Widget/Detail/IncidentDetail.php
@@ -105,19 +105,10 @@ class IncidentDetail extends BaseHtmlElement
 
     protected function createObjectTag(): array
     {
-        $tags = [];
-        foreach ($this->incident->object->object_extra_tag as $extraTag) {
-            $tags[] = Table::row([$extraTag->tag, $extraTag->value]);
-        }
-
-        if (! $tags) {
-            return $tags;
-        }
-
         return [
             new HtmlElement('h2', null, new Text(t('Object Tags'))),
             (new Table())
-                ->addHtml(...$tags)
+                ->addHtml()
                 ->addAttributes(['class' => 'object-tags-table'])
         ];
     }
```

</p>
</details>

oxzi commented on doc/20-HTTP-API.md:
This one is still open. I would like to update the documentation if everything else looks fine.

oxzi commented on internal/config/source.go:
```diff
@@ -18,16 +14,6 @@ type Source struct {
 	Name string `db:"name"`
 
 	ListenerPasswordHash types.String `db:"listener_password_hash"`
```
Done.

oxzi commented on internal/rule/rule.go:
```diff
@@ -16,21 +15,21 @@ type Rule struct {
+	// }
 
 	return nil
 }
```
This was a leftover. Now the template method is a noop.

oxzi commented on internal/object/objects.go:
```diff
@@ -99,15 +98,7 @@ func restoreObjectsFromQuery(ctx context.Context, db *database.DB, query string,
-						return errors.Wrap(err, "cannot restore objects extra tags")
-					}
-
+	
```
:ghost:

Removed and grepped the codebase for other occurrences - there were none. BTW, GoLand is now able to both show and remove the tab. Impressive.

sukhwinder33445 commented on schema/pgsql/upgrades/001.sql:
```diff
@@ -0,0 +1,20 @@
+  Drop COLUMN icinga2_ca_pem,
+  Drop COLUMN icinga2_common_name,
+  Drop COLUMN icinga2_insecure_tls,
+  ALTER COLUMN listener_password_hash SET NOT NULL;
```
This column was previously nullable, and I already have an entry with a null value. This means that I cannot perform this upgrade. The following is missing:
```sql
UPDATE source SET listener_password_hash = '$2_$%' WHERE listener_password_hash IS NULL;
```

oxzi commented on internal/event/event.go:
```diff
@@ -20,46 +24,34 @@ var ErrSuperfluousMuteUnmuteEvent = errors.New("ignoring superfluous (un)mute ev
-	ID int64 `json:"-"`
+	u, err := url.Parse(e.URL)
+	if err != nil || u.Scheme == "" {
+		e.URL = icingaWebBaseUrl + e.URL
```
I like the idea, but this would require further communication with web, as you have written. Unsure if this fits in the schedule.

oxzi commented on internal/incident/incident.go:
```diff
@@ -395,29 +394,30 @@ func (i *Incident) handleMuteUnmute(ctx context.Context, tx *sqlx.Tx, ev *event.
+			return fmt.Errorf("cannot apply unknown rule %d", ruleIdInt)
+		}
 
+		if _, ok := i.Rules[r.ID]; !ok {
```
Good catch. I have added a check.

yhabteab commented on internal/config/runtime.go:
```diff
@@ -68,9 +62,10 @@ type ConfigSet struct {
 	Sources          map[int64]*Source
 
+	RuleSet // RuleSet contains the currently loaded rules and their version.
+
```
The point is that I didn't want to bloat the `ConfigSet` struct with even mor fields that are actually all `rule` related, thus I've used `RuleSet` to group all the fields concerning event rules. -

oxzi commented on internal/event/event.go:
```diff
@@ -73,51 +65,28 @@ func (e *Event) Validate() error {
-		return fmt.Errorf("invalid event: unsupported event type %q", e.Type)
+	if e.Type == baseEv.TypeUnknown {
+		return fmt.Errorf("invalid event: unsupported or empty event type %q", e.Type)
 	}
```
Changed the error according to @julianbrost's suggestion.

oxzi replied:
Changed.

julianbrost commented on schema/pgsql/upgrades/001.sql:
I'd just rename this to something like `0.2.0-source-rules.sql` similar to how we did with the other recent PRs. This just avoids name conflicts altogether and all schema upgrades will be combined into a single file before the release anyways, see https://github.com/Icinga/icinga-notifications/pull/344#discussion_r2472319727 for more details.

oxzi replied:
Valid point, done.

julianbrost commented on internal/incident/incidents_test.go:
```diff
@@ -22,10 +23,10 @@ func TestLoadOpenIncidents(t *testing.T) {
+	source := &config.Source{
+		Type:                 "notifications",
+		Name:                 "Icinga Notifications",
+		ListenerPasswordHash: types.MakeString("$2y$"), // Needed to pass the database constraint.
```
> Needed to pass the database constraint.

That constraint no longer exists.

oxzi commented on doc/20-HTTP-API.md:
```diff
@@ -35,7 +35,9 @@ curl -v -u 'source-2:insecureinsecure' -d '@-' 'http://localhost:5680/process-ev
+  "message": "Something went somewhere very wrong.",
+  "rule_version": "23",
+  "rule_ids": ["0"],
 }
```
Oops, changed.

oxzi commented on schema/pgsql/schema.sql:
```diff
@@ -239,37 +239,23 @@ CREATE INDEX idx_timeperiod_entry_changed_at ON timeperiod_entry(changed_at);
-    -- to the Listener. The username will be "source-${id}", allowing early verification.
+    -- This column is required to limit API access for incoming connections to the Listener.
+    -- The username will be "source-${id}", allowing early verification.
     listener_password_hash text,
```
Removed.

julianbrost commented on go.mod:
```diff
@@ -2,13 +2,14 @@ module github.com/icinga/icinga-notifications
 	github.com/google/uuid v1.6.0
-	github.com/hashicorp/golang-lru/v2 v2.0.7
-	github.com/icinga/icinga-go-library v0.7.2
+	github.com/icinga/icinga-go-library v0.7.3-0.20251022120618-6600889adc38
```
That commit (https://github.com/Icinga/icinga-go-library/commit/6600889adc38ffe2088948411c852cb387b5762f) is included in the `main` branch of that repo, so it'll be included in the next release. Pinning a commit should be fine for the meantime, but for a tagged version in this repo, a tagged version from that repo should be used (I'll create an issue so that we won't forget).

julianbrost replied:
> I'll create an issue so that we won't forget

Done: #347

nilmerg commented:
Since this drops the icinga2 source and Icinga DB only transmits customvars as extra tags, wouldn't it be a good idea to drop all extra tags during migration? Otherwise old extra tags remain and won't be deleted. Or is this a general limitation since it also means customvars aren't deleted?
